cheatsheets/DotNet_Security_Cheat_Sheet.md

```diff
@@ -493,13 +741,197 @@ After .NET Core 2.0 it is possible to automatically generate and verify the anti
+
+```csharp
+System.IO.FileInfo
+```    
+
+`FileInfo` objects that reference files actually on the server can when deserialized, change the properties of those files e.g. to read-only, creating a potential denial of service attack.
+
+Even if you have limited the types that can be deserialised remember that some types have properties that are risky. `System.ComponentModel.DataAnnotations.ValidationException`, for example has a property `Value` of type `Object`. if this type is the type allowed for deserialization then an attacker can set the `Value` property to any object type they choose.
+
+Attackers should be prevented from steering the type that will be instantiated. If this is possible then even `DataContractSerializer` or `XmlSerializer` can be subverted e.g.
+
+```csharp
+// Action below is dangerous if the attacker can change the data in the database
+var typename = GetTransactionTypeFromDatabase();  
+
+var serializer = new DataContractJsonSerializer(Type.GetType(typename));
+
+var obj = serializer.ReadObject(ms);
+```    
+
+Execution can occur within certain .Net types during deserialization. Creating a control such as the one shown below is ineffective.
+
+```csharp
+var suspectObject = myBinaryFormatter.Deserialize(untrustedData);
+
+//Check below is too late! Execution may have already occurred.
+if (suspectObject is SomeDangerousObjectType) 
+{
+    //generate warnings and dispose of suspectObject
+}
+```    
+
+For `BinaryFormatter` and `JSON.Net` it is possible to create a safer form of white list control useing a custom `SerializationBinder`.
+
+Try to keep up-to-date on known .Net insecure deserialization gadgets and pay special attention where such types can be created by your deserialization processes. **A deserializer can only instantiate types that it knows about**. 
+
+Try to keep any code that might create potential gagdets separate from any code that has internet connectivity. As an example `System.Windows.Data.ObjectDataProvider` used in WPF applications is a known gadget that allows arbitrary method invocation. It would be risky to have this a reference to this assembly in a REST service project that deserializes untrusted data.
+
+### Known .NET RCE Gadgets
+
+- `System.Configuration.Install.AssemblyInstaller`
+- `System.Activities.Presentation.WorkflowDesigner`
+- `System.Windows.ResourceDictionary`
+- `System.Windows.Data.ObjectDataProvider`
+- `System.Windows.Forms.BindingSource`
+- `Microsoft.Exchange.Management.SystemManager.WinForms.ExchangeSettingsProvider`
+- `System.Data.DataViewManager, System.Xml.XmlDocument/XmlDataDocument`
+- `System.Management.Automation.PSObject`
+
+More information can be found here: [Deserialization Cheat Sheet](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Deserialization_Cheat_Sheet.md)
```

Comment:
Fix link to correct section https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Deserialization_Cheat_Sheet.md#net-csharp

Reply:
Done